File: apps/web/src/middleware.ts
import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';

const isDev = process.env.NODE_ENV === 'development';

/**
 * CSP directives in report-only mode.
 *
 * Next.js injects inline scripts for hydration and uses style-loader with
 * inline styles in development, so we allow 'unsafe-inline' for both to
 * avoid breaking the UI.  Once a nonce strategy is in place these can be
 * tightened.
 *
 * 'unsafe-eval' is only included in development (Next.js hot-reload needs it).
 * connect-src uses localhost/api:8000 in dev; in prod it uses NEXT_PUBLIC_API_URL.
 */
const apiOrigin = process.env.NEXT_PUBLIC_API_URL ?? 'http://localhost:8000';

const connectSrc = isDev
    ? "connect-src 'self' http://localhost:8000 http://api:8000 https://accounts.google.com/gsi/"
    : `connect-src 'self' ${apiOrigin} https://accounts.google.com/gsi/`;

const scriptSrc = isDev
    ? "script-src 'self' 'unsafe-inline' 'unsafe-eval' https://accounts.google.com/gsi/client"
    : "script-src 'self' 'unsafe-inline' https://accounts.google.com/gsi/client";

const CSP_DIRECTIVES = [
    "default-src 'self'",
    scriptSrc,
    "style-src 'self' 'unsafe-inline' https://accounts.google.com/gsi/style",
    "img-src 'self' data: https://via.placeholder.com https://images.unsplash.com",
    "font-src 'self' data:",
    connectSrc,
    "frame-src 'self' https://accounts.google.com https://accounts.google.com/gsi/",
    "frame-ancestors 'none'",
    "base-uri 'self'",
    "form-action 'self'",
].join('; ');

/**
 * Middleware — sets x-pathname header so server components can read the
 * current path via `headers().get('x-pathname')`.
 *
 * This is required because Next.js App Router server components cannot access
 * the current URL directly.  The layout enforcement gate in /app/layout.tsx
 * uses x-pathname to detect billing-bypass paths (/app/planes, /app/servicios)
 * and avoid redirect loops.
 *
 * Also sets Content-Security-Policy-Report-Only for observability (Phase 2C).
 */
// Routes where Google Identity Services OAuth popup is used.
const GOOGLE_SIGNIN_PATHS = ['/entrar', '/registrarse'];

export function middleware(request: NextRequest) {
    const response = NextResponse.next();
    response.headers.set('x-pathname', request.nextUrl.pathname);
    response.headers.set('Content-Security-Policy-Report-Only', CSP_DIRECTIVES);

    // Allow the Google Sign-In popup to postMessage back to this page.
    // Only needed on pages that initiate the Google OAuth flow.
    const pathname = request.nextUrl.pathname;
    if (GOOGLE_SIGNIN_PATHS.some((p) => pathname === p || pathname.startsWith(p + '/'))) {
        response.headers.set('Cross-Origin-Opener-Policy', 'same-origin-allow-popups');
    }

    return response;
}

export const config = {
    matcher: [
        /*
         * Match all paths except static files and Next.js internals.
         * This keeps the middleware lightweight and fast.
         */
        '/((?!_next/static|_next/image|favicon.ico|logo/).*)',
    ],
};

File: services/api/src/config/middleware.py
"""
Content-Security-Policy-Report-Only middleware for Django API responses.

Adds a conservative CSP in report-only mode so violations are observable
in the browser console without blocking anything.  This covers the DRF
browsable API and Django Admin; the main UI is covered by the Next.js
middleware.

Phase 2C — observation only.  Switch to ``Content-Security-Policy`` (enforce)
once violations have been triaged.
"""
from django.http import HttpRequest, HttpResponse


# The API mostly returns JSON, but DRF browsable API and Django admin serve HTML.
# Keep directives lenient; this is report-only.
_CSP_POLICY = "; ".join([
    "default-src 'self'",
    "script-src 'self' 'unsafe-inline'",
    "style-src 'self' 'unsafe-inline'",
    "img-src 'self' data:",
    "font-src 'self' data:",
    "connect-src 'self'",
    "frame-ancestors 'none'",
    "base-uri 'self'",
    "form-action 'self'",
])


class CSPReportOnlyMiddleware:
    """Append Content-Security-Policy-Report-Only to every response."""

    def __init__(self, get_response):
        self.get_response = get_response

    def __call__(self, request: HttpRequest) -> HttpResponse:
        response = self.get_response(request)
        response['Content-Security-Policy-Report-Only'] = _CSP_POLICY
        return response
